ccsdspy/utils.py

```diff
@@ -0,0 +1,66 @@
+"""Utils for the CCSDSPy package."""
+
+__author__ = "Daniel da Silva <mail@danieldasilva.org>"
+
+from io import BytesIO
+import warnings
+
+import numpy as np
+
+
+def split_by_apid(mixed_file, valid_apids=None):
+    """Split a stream of mixed APIDs into seperate streams by APID. 
+    
+    Parameters
+    ----------
+    mixed_file: str
+       Path to file on the local file system, or file-like object
+    valid_apids: list of int, None
+       Optional list of valid APIDs. If specified, warning will be issued when
+       an APID is encountered outside this list. 
+
+    Returns
+    -------
+    Dictionary mapping integer apid number to BytesIO instance with the file
+    pointer at the beginning of the file.
+    """
+    if hasattr(mixed_file, 'read'):
+        file_bytes = np.frombuffer(mixed_file.read(), 'u1')
+    else:
+        file_bytes = np.fromfile(mixed_file, 'u1')
+
+    offset = 0
+    stream_by_apid = {}
+    
+    while offset < len(file_bytes):
+        packet_nbytes = file_bytes[offset + 4] * 256 + file_bytes[offset + 5] + 7
+            
+        apid = np.array(
+            [file_bytes[offset], file_bytes[offset+1]],
+            dtype=np.uint8
+        )
+        apid.dtype = '>u2'
+        apid = apid[0] 
+        apid &= 0x07FF
+                
+        if valid_apids is not None and apid not in valid_apids:
+            warnings.warn('Found unknown APID %d' % apid)
```

Comment:
This is technically fine, but I think an f-string is generally preferred now (although it does require Python 3.6 or higher I believe). So maybe replace with `warnings.warn(f'Found unknown APID {apid}')`. I think there are few spots in the new code that do this. 

Some packages, SunPy for example dictate not to use the old % formatting convention in their [style guides](https://docs.sunpy.org/en/latest/dev_guide/contents/code_standards.html#language-standard).

Reply:
Technically the last release still supported 2.7!... but it's probably time we can cut those people off. This is a good idea and I'll move towards f-strings